app/features/notes_generator/tools.py

```diff
@@ -0,0 +1,199 @@
+
+def read_text_file(file_path):
+    # Get the directory containing the script file
+    script_dir = os.path.dirname(os.path.abspath(__file__))
+
+    # Combine the script directory with the relative file path
+    absolute_file_path = os.path.join(script_dir, file_path)
+    
+    with open(absolute_file_path, 'r') as file:
+        return file.read()
+    
+class NotesGenerator:
+    def __init__(self, args=None, vectorstore_class=Chroma, prompt=None, embedding_model=None, model=None, parser=None, verbose=False):
+        default_config = {
+            "model": GoogleGenerativeAI(model="gemini-1.5-flash"),
+            "embedding_model": GoogleGenerativeAIEmbeddings(model='models/embedding-001'),
+            "parser": JsonOutputParser(pydantic_object=NotesOutput),
+            "prompt": read_text_file("prompt/notes-generator-prompt.txt"),
+            "vectorstore_class": Chroma
+        }
+
+        self.prompt = prompt or default_config["prompt"]
+        self.model = model or default_config["model"]
+        self.parser = parser or default_config["parser"]
+        self.embedding_model = embedding_model or default_config["embedding_model"]
+        self.example = read_text_file("prompt/example.txt")
+
+        self.vectorstore_class = vectorstore_class or default_config["vectorstore_class"]
+        self.vectorstore, self.retriever, self.runner = None, None, None
+        self.args = args
+        self.verbose = verbose
+
+        if vectorstore_class is None: raise ValueError("Vectorstore must be provided")
+        if args.orientation is None: raise ValueError("Orientattion must be provided")
+        if args.nb_columns is None: raise ValueError("nb_columns must be provided")
+        if args.details is None: raise ValueError("details about the notes must be provided")
+        if args.topic is None: raise ValueError("topic must be provided")
+
+
+    def compile(self, documents: List[Document]):
+        prompt = PromptTemplate(
+            template=self.prompt,
+            input_variables=["attribute_collection"],
+            partial_variables={"format_instructions": self.parser.get_format_instructions(),
+                               "example": self.example 
+                               }
+        )
+
+        if self.runner is None:
+            logger.info(f"Creating vectorstore from {len(documents)} documents") if self.verbose else None
+            self.vectorstore = self.vectorstore_class.from_documents(documents, self.embedding_model)
+            logger.info(f"Vectorstore created") if self.verbose else None
+
+            self.retriever = self.vectorstore.as_retriever()
+            logger.info(f"Retriever created successfully") if self.verbose else None
+
+            self.runner = RunnableParallel(
+                {"context": self.retriever,
+                 "attribute_collection": RunnablePassthrough()
+                }
+            )
+
+        chain = self.runner | prompt | self.model | self.parser
```

Comment:
Make sure the time that it takes in order to return a response is lower than 30 seconds. If not, you can use other strategies for improving it.

Reply:
It takes less than 13 seconds with 2 URLs, we need to test it with more files as input.